### PR TITLE
Checkpoint ledger arithmetic and catchup range arithmetic cleanup

### PR DESCRIPTION
**Pull request by graydon**
This is somewhat involved in terms of the quantity of cleanup -- far more than I was expecting at first! -- but ultimately only fixes two "bugs" (that you might not even consider as bugs):

  - The "bug" that motivated all this: running catchup on a range like 66/4 would attempt to catchup by applying checkpoint 63 then replaying 3 ledgers: 64, 65, 66. This might or might not be a "bug" depending on how you look at it, but it certainly doesn't do two of the most likely things the user might have wanted: it doesn't replay 4 ledgers, and it doesn't replay ledger 63 (which is 66+1-4). So now this is fixed! Now we'll replay ledgers 1..66 in that case -- to be sure we span at least 4 ledgers before 66 -- and the cases where we'll apply buckets are those where it happens _strictly before_ the range the user requested. This fix starts with removing an incorrect special case in `computeCatchupledgers` in 56dbb965e0b6358171452b1d2329a47a99ad1a3f but that function is subsequently entirely rewritten and then replaced in 06aef10726db634d8316c719c59a9dd6b47938d8 and 8acefa2202fdecb9c5369ad1639534b2d28f102b

  - The `CatchupManagerImpl::processLedger` function was a little too dependent on the catchup trigger window being exactly one ledger past the checkpoint at the start of the ledger buffer: if that's ever changed in the future (eg. due to fixing https://github.com/stellar/stellar-core/issues/2455) we might introduce a gap in the ledger buffer. So that entire function was rewritten a bit with an eye to clarity (and eliminating that potential future "bug") in 220ad10cb942cff3f64f52830842c6fc3e70a8f6

The remainder of the PR is just "cleanup", which broadly falls into these categories:

  - Removing the concept of "current ledger" (one past LCL) from the `HistoryManager` interface entirely. This has been there for years and has always been a source of confusion.
  - Adding a bunch of helper methods to `HistoryManager` with semantically-more-meaningful names than `nextCheckpointLedger` and `prevCheckpointLedger` (which were related to "current ledger" anyway), and moving all code that used those two methods over to the new more-meaningfully-named methods.
  - Splitting the `CatchupRange` type from a combined (and extremely confusing) hybrid representation of a bucket-apply checkpoint and a ledger-apply range to a class that separately represents those elements, and improving/simplifying/documenting the logic that constructs it.
  - Changing `LedgerRange`, `CheckpointRange` and `CatchupRange` to deal explicitly with empty ledger-apply ranges in catchup, switching them from closed `[lo,hi]` ranges to half-open `[lo,lo+count)` ranges that support a count of zero.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [x] N/A If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/catchup/DownloadApplyTxsWork.cpp [hunk 1]
```diff
@@ -46,10 +46,14 @@ DownloadApplyTxsWork::yieldMoreWork()
...
-    auto high = std::min(mCheckpointToQueue, mRange.mLast);
-    auto apply = std::make_shared<ApplyCheckpointWork>(mApp, mDownloadDir,
-                                                       LedgerRange{low, high});
+    auto start = hm.firstLedgerInCheckpointContaining(mCheckpointToQueue);
+    auto limit =
+        std::min(mRange.limit(),
+                 hm.firstLedgerAfterCheckpointContaining(mCheckpointToQueue));
+    assert(limit >= start);
+    auto count = limit - start;
+    auto apply = std::make_shared<ApplyCheckpointWork>(
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/DownloadApplyTxsWork.cpp]: Oh, I think I edited this hunk before I added the last() and inclusive() methods to range. Fixed.

File: src/catchup/DownloadApplyTxsWork.cpp [hunk 2]
```diff
@@ -120,7 +123,7 @@ bool
 DownloadApplyTxsWork::hasNext() const
 {
     auto last =
-        mApp.getHistoryManager().checkpointContainingLedger(mRange.mLast);
+        mApp.getHistoryManager().checkpointContainingLedger(mRange.last());
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/DownloadApplyTxsWork.cpp]: True! Added a guard.

File: src/catchup/ApplyCheckpointWork.cpp [hunk 3]
```diff
@@ -286,7 +285,8 @@ ApplyCheckpointWork::onRun()
         }
 
         auto const& lm = mApp.getLedgerManager();
-        auto done = lm.getLastClosedLedgerNum() == mLedgerRange.mLast;
+        auto done = (mLedgerRange.mCount == 0 ||
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/ApplyCheckpointWork.cpp]: Fixed.

File: src/catchup/CatchupRange.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+calculateCatchupRange(uint32_t lcl, CatchupConfiguration const& cfg,
+                      HistoryManager const& hm)
+{
+    checkCatchupPreconditions(lcl, cfg);
+    const uint32_t init = LedgerManager::GENESIS_LEDGER_SEQ;
+
+    // Case 1: replay from LCL as we're already past genesis.
+    if (lcl > init)
+    {
+        LedgerRange replay(lcl + 1, cfg.toLedger() - lcl);
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/CatchupRange.cpp]: Good call! Done (though I called it `fullReplayCount`).

File: src/catchup/CatchupRange.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        // CLOG(TRACE, "History") << "CatchupRange: case 1 (lcl is past
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/CatchupRange.cpp]: Done.

File: src/catchup/CatchupRange.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    // Case 3: special case of buckets only, no replay; only
+    // possible when targeting the exact end of a checkpoint.
+    if (cfg.count() == 0 && hm.isLastLedgerInCheckpoint(cfg.toLedger()))
+    {
+        uint32_t applyBucketsOnlyAt = cfg.toLedger();
+        // CLOG(TRACE, "History") << "CatchupRange: case 3 (checkpoint
+        // boundary)";
+        return CatchupRange(applyBucketsOnlyAt);
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/CatchupRange.cpp]: Done.

File: src/catchup/CatchupRange.h [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+  public:
+    // Return a LedgerRange spanning both the apply-buckets phase (if it
+    // exists) and the replay phase.
+    LedgerRange
+    getFullRangeIncludingBucketApply() const
+    {
+        if (mApplyBuckets)
+        {
+            assert(mReplayRange.mCount == 0 ||
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/CatchupRange.h]: Ok, removed.

File: src/catchup/VerifyLedgerChainWork.cpp [hunk 8]
```diff
@@ -104,9 +107,12 @@ VerifyLedgerChainWork::getStatus() const
 {
     if (!isDone() && !isAborting())
     {
-        std::string task = "verifying checkpoint";
-        return fmtProgress(mApp, task, mRange.mFirst, mRange.mLast,
-                           (mRange.mLast - mCurrCheckpoint));
+        if (mRange.mCount != 0)
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/VerifyLedgerChainWork.cpp]: Fixed.

File: src/ledger/CheckpointRange.h [hunk 9]
```diff
@@ -4,31 +4,65 @@
...
 
-    CheckpointRange(uint32_t first, uint32_t last, uint32_t frequency);
+    CheckpointRange(uint32_t first, uint32_t count, uint32_t frequency);
+    static CheckpointRange
+    inclusive(uint32_t first, uint32_t last, uint32_t frequency)
+    {
+        // CheckpointRange is half-open: in exchange for being able to represent
+        // empty ranges, it can't represent ranges that include UINT32_MAX.
+        assert(last < std::numeric_limits<int32_t>::max());
+        return CheckpointRange(first, last - first + 1, frequency);
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/CheckpointRange.h]: Good catch! Luckily nothing is using this code except the mostly-dead command that fetches recent qsets (residue from an earlier attempt at quorum safety checking). I should probably kill that code off eventually, but anyway, fixed this part for now!

File: src/catchup/simulation/HistoryArchiveStream.cpp [hunk 10]
```diff
@@ -190,7 +202,7 @@ HistoryArchiveStream::getNextLedger(LedgerHeaderHistoryEntry& header,
                                     TransactionHistoryEntry& transaction,
                                     TransactionHistoryResultEntry& result)
 {
-    if (mHeaderHistory.header.ledgerSeq < mRange.mLast)
+    if (mHeaderHistory.header.ledgerSeq < mRange.limit())
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/catchup/simulation/HistoryArchiveStream.cpp]: Correct! Fixed.

**Comment by graydon** [comment]
Rebased and addressed all comments.

**Comment by MonsieurNicolas** [comment]
r+ fe1a464c8cc6e60ee83799b9e47aadd3276426cf

**Pull request merged**